sycl/doc/extensions/ExtensionMechanism/SYCL_INTEL_extension_api.asciidoc

```diff
@@ -0,0 +1,620 @@
+defined in the core SYCL specification, or through a combination of the two.
+
+New types or free functions for Khronos ratified extensions are defined in the
+namespace `sycl::khr::<extensionname>`.  For example, `sycl::khr::fancy` could
+be the namespace for a Khronos extension named "fancy"
+
+If a vendor specific extension adds new types or free functions, the vendor is
+encouraged to define them in the namespace `sycl::ext::<vendorname>` and they
+are encouraged to add another namespace layer according to the name of the
+extension.  For example, `sycl::ext::acme::fancier` could be the namespace for
+an extension from the Acme vendor.  However, vendors may also choose to define
+new types and free functions in another top-level namespace that is outside of
+`sycl`.  This might be more appropriate, for example, when an extension
+integrates features from an existing non-SYCL API.  A vendor may not define new
+types or free functions underneath `sycl`, unless they are in
+`sycl::ext::<vendorname>`.
+
+[NOTE]
+====
+Vendors are discouraged from defining top level namespaces that start with the
+word "sycl" because we believe that application developers may want to use
+namespaces like this as namespace aliases.
+====
+
+Extensions may only add functionality to existing SYCL classes in a limited
+way.  When a Khronos ratified extension needs to add functionality to an
+existing class, it does so by adding a method named `khr()` to that class.  For
+example, an extension on the `device` class would add a method like this:
+
+[source,c++,NoName,linenums]
+----
+class device {
+  // ...
+  sycl::khr::device khr();
+};
+----
+
+The `khr()` method returns an object, and that object provides methods that
+are part of the extension.
+
+Likewise, a vendor specific extension may add functionality to an existing
+SYCL class by adding a method named `ext_<vendorname>()` (e.g. `ext_acme()`
+for the Acme vendor) like this:
+
+[source,c++,NoName,linenums]
+----
+class device {
+  // ...
+  sycl::ext::acme::device ext_acme();
+};
+----
+
+One motivation for this pattern is to reduce verbosity of application code
+that uses an extension and to facilitate application migration when an
+extension is promoted to the core SYCL specification.  Consider the following
+application code:
+
+[source,c++,NoName,linenums]
+----
+void foo(sycl::device dev) {
+  dev.ext_acme().fancy();
+};
+----
+
+If the extension "fancy" is later promoted to the core SYCL specification, the
+application need only remove the call to `ext_acme()` in order to migrate the
+application.
+
+If an extension requires {cpp} attributes, those attributes must be defined in an
+attribute namespace that corresponds to the regular namespace rules described
+above.  Thus, the attribute namespace `sycl` is reserved for {cpp} attributes
+used in the core SYCL specification, `sycl::khr` is reserved for Khronos
+ratified extensions, and `sycl::ext::<vendorname>` is reserved for vendor
+extensions.
```

Comment:
@gmlueck, I wonder how is that correlates with C++ which is used as a base for SYCL? If I understand correctly, C++ allows only one namespace within attribute name, [docs](https://en.cppreference.com/w/cpp/language/attributes)

Reply:
Good point.  This part of the spec should be rewritten as follows:

> Extensions may also add C++ attributes. The attribute namespace `sycl::` is reserved for attributes in the core SYCL specification and for Khronos ratified extensions. Vendor defined extensions should use a different attribute namespace.